externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/TMTensor/IR/TMTensorOps.cpp

```diff
@@ -84,6 +84,254 @@ OpFoldResult TMTensor::getDim(OpBuilder &builder, Location loc, Value v,
+            loc, min, max, steps,
+            [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+              SmallVector<Value> coords(localIVs);
+              coords.push_back(outerDim);
+              Value x = b.create<memref::LoadOp>(loc, weight, coords);
+              x = b.create<math::ExpOp>(loc, x);
+              b.create<memref::StoreOp>(loc, x, weight, coords);
+              b.create<scf::YieldOp>(loc);
+            });
+      });
+  // calculate sum(exp(weight), -1)
+  Value expWeightSum = b.create<memref::AllocOp>(
+      loc,
+      MemRefType::get(
+          SmallVector<int64_t>(weightSizes.begin(), weightSizes.end() - 1),
+          elementType),
+      SmallVector<Value>{weightDynSizes.begin(), weightDynSizes.end() - 1});
+  b.create<scf::ParallelOp>(
+      loc, SmallVector<Value>(queryRank - 1, zero),
+      SmallVector<Value>{weightDynSizes.begin(), weightDynSizes.end() - 1},
+      SmallVector<Value>(queryRank - 1, one),
+      [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+        b.create<memref::StoreOp>(loc, zeroF, expWeightSum, localIVs);
+      });
+  // Loop over all dims but -1
+  b.create<scf::ParallelOp>(
+      loc, SmallVector<Value>(queryRank - 1, zero),
+      SmallVector<Value>(weightDynSizes.begin(), weightDynSizes.end() - 1),
+      SmallVector<Value>(queryRank - 1, one),
+      [&](OpBuilder &b, Location loc, ValueRange outsideDims) {
+        // Sum over last dim
+        b.create<scf::ParallelOp>(
+            loc, zero, dim, one,
+            [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+              SmallVector<Value> coords(outsideDims);
+              coords.push_back(localIVs[0]);
+              Value x =
+                  b.create<memref::LoadOp>(loc, expWeightSum, outsideDims);
+              Value y = b.create<memref::LoadOp>(loc, weight, coords);
+              Value sum = b.create<arith::AddFOp>(loc, x, y);
+              b.create<memref::StoreOp>(loc, sum, expWeightSum, outsideDims);
+              b.create<scf::YieldOp>(loc);
+            });
+      });
+  // calculate exp(weight) / sum(exp(weight))
+  b.create<scf::ParallelOp>(
+      loc, SmallVector<Value>(queryRank - 1, zero),
+      SmallVector<Value>(weightDynSizes.begin(), weightDynSizes.end() - 1),
+      SmallVector<Value>(queryRank - 1, one),
+      [&](OpBuilder &b, Location loc, ValueRange localIVs) {
+        Value sum = b.create<memref::LoadOp>(loc, expWeightSum, localIVs);
+        b.create<scf::ParallelOp>(
+            loc, zero, dim, one,
+            [&](OpBuilder &b, Location loc, ValueRange outerDim) {
+              SmallVector<Value> coords(localIVs);
+              coords.push_back(outerDim[0]);
+              Value x = b.create<memref::LoadOp>(loc, weight, coords);
+              x = b.create<arith::DivFOp>(loc, x, sum);
+              b.create<memref::StoreOp>(loc, x, weight, coords);
+              b.create<scf::YieldOp>(loc);
+            });
+      });
```

Comment:
Similar comment here: this is an elementwise operation on `weight` so a single `ParallelOp` should be enough.

Reply:
I wrote it that way to avoid loading `sum` multiple times along the last dim, but if it's clearer this way then I'll rewrite it.